exchange/floors.go

```diff
@@ -0,0 +1,134 @@
+					glog.Errorf(errMsg)
+					rejections = append(rejections, errMsg)
+					continue
+				}
+				bidPrice = rate * bid.bid.Price
+			}
+			if bidFloor.bidFloor > bidPrice {
+				rejections = updateRejections(rejections, bidID, fmt.Sprintf("bid price value %.4f %s is less than bidFloor value %.4f %s for impression id %s bidder %s", bidPrice, bidFloor.bidFloorCur, bidFloor.bidFloor, bidFloor.bidFloorCur, bid.bid.ImpID, bidderName))
+				continue
+			}
+			eligibleBids = append(eligibleBids, bid)
+		}
+		seatBids[bidderName].bids = eligibleBids
+	}
+	return seatBids, rejections
+}
+
+// signalFloors function does singanlling of floors,
+// Internally validation of floors parameters and validation of rules is done,
+// Based on number of modelGroups and modelWeight, one model is selected and imp.bidfloor and imp.bidfloorcur is updated
+func signalFloors(r *AuctionRequest, floor floors.Floor, conversions currency.Conversions, responseDebugAllow bool) []error {
+
+	var errs []error
+	requestExt, err := r.BidRequestWrapper.GetRequestExt()
+	if err != nil {
+		errs = append(errs, err)
+		return errs
+	}
+	prebidExt := requestExt.GetPrebid()
+	if floor != nil && floor.IsFloorEnabled() && prebidExt.Floors.GetEnabled() && prebidExt.Floors != nil {
+		errs = floors.ModifyImpsWithFloors(prebidExt.Floors, r.BidRequestWrapper.BidRequest, conversions)
+		requestExt.SetPrebid(prebidExt)
+		err := r.BidRequestWrapper.RebuildRequest()
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+		if responseDebugAllow {
+			updatedBidReq, _ := json.Marshal(r.BidRequestWrapper.BidRequest)
+			//save updated request after floors signalling
+			r.UpdatedBidRequest = updatedBidReq
+		}
+	}
+	return errs
+}
+
+// enforceFloors function does floors enforcement
+func enforceFloors(r *AuctionRequest, seatBids map[openrtb_ext.BidderName]*pbsOrtbSeatBid, floor floors.Floor, conversions currency.Conversions, responseDebugAllow bool) (map[openrtb_ext.BidderName]*pbsOrtbSeatBid, []string) {
+
+	var rejections []string
+	requestExt, err := r.BidRequestWrapper.GetRequestExt()
+	if err != nil {
+		rejections = append(rejections, err.Error())
+		return seatBids, rejections
+	}
+	prebidExt := requestExt.GetPrebid()
+	if floor != nil && floor.IsFloorEnabled() && prebidExt.Floors.GetEnabled() {
+		if floors.ShouldEnforce(r.BidRequestWrapper.BidRequest, prebidExt.Floors, floor.GetEnforceRate(), rand.Intn) {
+			var enforceDealFloors bool
+			if prebidExt != nil && prebidExt.Floors != nil && prebidExt.Floors.Enforcement != nil && prebidExt.Floors.Enforcement.FloorDeals != nil {
+				enforceDealFloors = *prebidExt.Floors.Enforcement.FloorDeals && floor.EnforceDealFloor()
+			}
+			seatBids, rejections = enforceFloorToBids(r.BidRequestWrapper.BidRequest, seatBids, conversions, enforceDealFloors)
+		}
+		requestExt.SetPrebid(prebidExt)
+		err = r.BidRequestWrapper.RebuildRequest()
```

Comment:
`enforceFloors` is only called from `HoldAuction`. It's odd to see a change to the request happening here after it's already been sent to the bidder. I also find it confusing that `prebidExt` would be modified in `floors.ShouldEnforce`.

All code which modifies a request for a specific bidder is contained in `cleanOpenRTBRequest`. Could you please refactor this code to move this logic there? ... except for `enforceFloorToBids` as since that operates on the bids it needs to stay in this location.

Reply:
refer https://github.com/prebid/prebid-server/pull/2341#discussion_r958319263